sql: allow listing orphan dropped tables in crdb_internal.tables

knz opened the pull request:
The virtual table `crdb_internal.tables` purposefully lists all table
descriptors, including dropped descriptors, for debugging purposes.

Prior to this patch, an error was emitted when a dropped descriptor
didn't have a valid parent database. The situation should not cause an
error, however, as a dropped table descriptor can exist with no
parent, e.g. as a consequence of DROP DATABASE CASCADE.

This patch fixes the issue by listing the parent ID of a dropped table
in numeric form if the database descriptor cannot be found.

Fixes #18432.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/18572)
<!-- Reviewable:end -->

nvanbenschoten commented on pkg/sql/logictest/testdata/logic_test/drop_database:
```diff
@@ -22,6 +22,9 @@ DROP DATABASE "foo-bar" RESTRICT
 DROP DATABASE "foo-bar" CASCADE
 
+statement ok
+SELECT * FROM crdb_internal.tables WHERE name = 'foo-bar'
```
Is it possible to create a test where we see the `[<ParentID>]` column?

nvanbenschoten commented on pkg/sql/information_schema.go:
```diff
@@ -743,6 +743,11 @@ func forEachTableDescWithTableLookupInternal(
 		if table, ok := desc.(*sqlbase.TableDescriptor); ok && !table.Dropped() {
 			dbName, ok := dbIDsToName[table.GetParentID()]
 			if !ok {
+				// Contrary from `crdb_internal.tables`, where dropped tables
```
`s/Contrary from/Contrary to/`

nvanbenschoten replied:
Also, could you give a reason for the difference in this comment? Right now you just state that a difference exists

vivekmenezes commented on pkg/sql/crdb_internal.go:
```diff
@@ -193,9 +193,13 @@ CREATE TABLE crdb_internal.tables (
+				// a database is dropped with CASCADE, and someone queries
+				// this virtual table before the dropped table descriptors are
+				// effectively deleted.
+				dbName = fmt.Sprintf("[%d]", table.GetParentID())
```
Is this visible in a test?

knz committed:
sql: allow listing orphan dropped tables in crdb_internal.tables

The virtual table `crdb_internal.tables` purposefully lists all table
descriptors, including dropped descriptors, for debugging purposes.

Prior to this patch, an error was emitted when a dropped descriptor
didn't have a valid parent database. The situation should not cause an
error, however, as a dropped table descriptor can exist with no
parent, e.g. as a consequence of DROP DATABASE CASCADE.

This patch fixes the issue by listing the parent ID of a dropped table
in numeric form if the database descriptor cannot be found.

knz commented:
Reviewed 2 of 3 files at r1.
Review status: 1 of 3 files reviewed at latest revision, 3 unresolved discussions.

---

*[pkg/sql/crdb_internal.go, line 202 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/18572#-KuMIC0shlQBL6fzc7LZ-r1-202:-KuQ1kJLHjX-1uAXpbg0:b-cdw158) ([raw file](https://github.com/cockroachdb/cockroach/blob/674113912756dc5309922734a84ec96069e7d31a/pkg/sql/crdb_internal.go#L202)):*
<details><summary><i>Previously, vivekmenezes wrote…</i></summary><blockquote>

Is this visible in a test?
</blockquote></details>

Now it is, added to `drop_database`

---

*[pkg/sql/information_schema.go, line 746 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/18572#-KuMIC0shlQBL6fzc7L_-r1-746:-KuQ22K28PiQmj7fxHUR:bnhjw4l) ([raw file](https://github.com/cockroachdb/cockroach/blob/674113912756dc5309922734a84ec96069e7d31a/pkg/sql/information_schema.go#L746)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Also, could you give a reason for the difference in this comment? Right now you just state that a difference exists
</blockquote></details>

Clarified in the comment.

---

*[pkg/sql/logictest/testdata/logic_test/drop_database, line 26 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/18572#-KuMIC0shlQBL6fzc7La-r1-26:-KuQ25Elhf2gKdjvK6nM:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/674113912756dc5309922734a84ec96069e7d31a/pkg/sql/logictest/testdata/logic_test/drop_database#L26)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Is it possible to create a test where we see the `[<ParentID>]` column?
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/18572)*
<!-- Sent from Reviewable.io -->

nvanbenschoten commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 1 of 2 files at r2.
Review status: 2 of 3 files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---

*[pkg/sql/information_schema.go, line 746 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/18572#-KuMIC0shlQBL6fzc7L_-r1-746:-KuQA5OoD4reRqEDxODJ:b-ucfknh) ([raw file](https://github.com/cockroachdb/cockroach/blob/674113912756dc5309922734a84ec96069e7d31a/pkg/sql/information_schema.go#L746)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

Clarified in the comment.
</blockquote></details>

Thanks!

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/18572#-:-KuQAB7GhEDmoDbEN5ah:bnfp4nl)*
<!-- Sent from Reviewable.io -->
